timesketch/lib/llms/features/llm_forensic_report.py

```diff
@@ -0,0 +1,304 @@
+            raise IOError("Error reading LLM prompt file.") from e
+
+        if "<EVENTS_JSON>" not in prompt_template:
+            logger.error("Prompt template is missing the <EVENTS_JSON> placeholder")
+            raise ValueError(
+                "LLM forensic report prompt template is missing the "
+                "required <EVENTS_JSON> placeholder."
+            )
+
+        prompt_text = prompt_template.replace("<EVENTS_JSON>", json.dumps(events_dict))
+        return prompt_text
+
+    def _run_timesketch_query(
+        self,
+        sketch: Sketch,
+        query_string: str = "*",
+        query_filter: Optional[dict] = None,
+        id_list: Optional[list] = None,
+        datastore: Optional[OpenSearch] = None,
+        timeline_ids: Optional[list] = None,
+    ) -> pd.DataFrame:
+        """Runs a timesketch query and returns results as a DataFrame.
+        Args:
+            sketch: The Sketch object to query.
+            query_string: Search query string.
+            query_filter: Dictionary with filter parameters.
+            id_list: List of event IDs to retrieve.
+            datastore: OpenSearch instance for querying.
+            timeline_ids: List of timeline IDs to query.
+        Returns:
+            pd.DataFrame: DataFrame containing query results.
+        Raises:
+            ValueError: If datastore is not provided or no valid indices are found.
+        """
+        if datastore is None:
+            raise ValueError("Datastore must be provided.")
+
+        if not query_filter:
+            query_filter = {}
+
+        if id_list:
+            id_query = " OR ".join([f'_id:"{event_id}"' for event_id in id_list])
+            query_string = id_query
+
+        all_indices = list({t.searchindex.index_name for t in sketch.timelines})
+        indices_from_filter = query_filter.get("indices", all_indices)
+
+        if "_all" in indices_from_filter:
+            indices_from_filter = all_indices
+
+        indices, timeline_ids = utils.get_validated_indices(indices_from_filter, sketch)
+
+        if not indices:
+            raise ValueError(
+                "No valid search indices were found to perform the search on."
+            )
+
+        result = datastore.search(
+            sketch_id=sketch.id,
+            query_string=query_string,
+            query_filter=query_filter,
+            query_dsl="",
+            indices=indices,
+            timeline_ids=timeline_ids,
+        )
+
+        return export.query_results_to_dataframe(result, sketch)
+
+    def generate_prompt(self, sketch: Sketch, **kwargs: Any) -> str:
+        """Generates the forensic report prompt based on events from a query.
+        Args:
+            sketch: The Sketch object containing events to analyze.
+            **kwargs: Additional arguments including:
+                - form: Form data containing query and filter information.
+                - datastore: OpenSearch instance for querying.
```

Comment:
On the backend, you actually don't need the datastore object to query. You only need the sketch_id and can then use the wrapper I mentioned above.

Reply:
Which wrapper "mentioned above" are you referring to? I can't seem to find it